### PR TITLE
Revise and simplify CFA for `typeof` check expressions

### LINKED ISSUE #49316: Nightly - `typeof x == "function"` does not always work as typeguard to remove a Record type from a union
# Bug Report

### 🔎 Search Terms

* typeof function type guard

### 🕗 Version & Regression Information

- This changed between versions 4.7 and 4.8 dev

### ⏯ Playground Link

[Minimal reproduction](https://www.typescriptlang.org/play?ts=4.8.0-dev.20220530#code/KYDwDg9gTgLgBAMwK4DsDGMCWEVzThTAcySmAGUZpgAecuUGYFAEwGc4IAjAK2AwB8ACgBQcOGRZI0wKAC44QoQEo4AXgFwAbhEwtVAHzgAlftBY0A1sACeEBHHIAaRao3bdLASNUBvMXAANsDwUBAQMKZSMvKu6po6eiIBmA5CMDZgwPYSwNGy6mpqcADkyOhYOCV+AeJhEVHSBcWSTVABAL4iHUA)

### 💻 Code

```ts
export function configureStore<S extends object>(
  reducer: (() => void) | Record<keyof S, () => void>
) {
  let rootReducer: () => void

  if (typeof reducer === 'function') {
    rootReducer = reducer
/*
Type '(() => void) | (Record<keyof S, () => void> & Function)' is not assignable to type '() => void'.
  Type 'Record<keyof S, () => void> & Function' is not assignable to type '() => void'.
    Type 'Record<keyof S, () => void> & Function' provides no match for the signature '(): void'.(2322)
*/
  }
}
```

### 🙁 Actual behavior

Here the `reducer` is a union between a function and a Record with a yet-unresolved key generic. In previous versions, the `  if (typeof reducer === 'function') {` would have narrowed the type down to the function type, but now it ends up with the function type OR `Record<keyof S, () => void> & Function` and cannot assign the result to the variable.

### 🙂 Expected behavior

No error. Our record is not a function.

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #49316 and several similar issues seen in #49343.

In #49119 we improved CFA to be more correct for intersections and `typeof x === "object"` checks. This introduced some unintended differences in classification of objects and functions that we didn't have coverage for in our regular test suites. This PR fixes those differences. The PR also simplifies our `typeof` narrowing logic, removing over 100 lines of code from `checker.ts`.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot pack this
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot perf test faster

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot perf test faster

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -25164,87 +25138,30 @@ namespace ts {
...
-                            return isTypeSubtypeOf(c, t) ? c : neverType;
-                        }
-                        if (c === globalFunctionType && isTypeSubtypeOf(c, t)) {
-                            return c;
-                        }
-                        return getIntersectionType([t, c]);
-                    });
-                });
+            function narrowTypeByTypeFacts(type: Type, impliedType: Type, facts: TypeFacts) {
+                return mapType(type, t =>
```
Comment (Andarist) [inline src/compiler/checker.ts:25142]: Just a note from a bystander - I know that nested ternaries are pretty common in this codebase but they are making debugging & stepping through the code much harder.

Comment (ahejlsberg) [reply src/compiler/checker.ts:25142]: I hear you. But I much prefer the succinctness of the more functional style, and when you're looking at ternaries you sort of implicitly know that you don't have to worry about assignments and side effects.

**Comment by ahejlsberg** [comment]
All tests are clean and performance is unaffected. This should be ready to merge.

**Review by RyanCavanaugh — APPROVED** [review]

**Pull request merged**